p2p/base_reactor.go

```diff
@@ -41,25 +46,151 @@ type Reactor interface {
 	// Receive is called by the switch when an envelope is received from any connected
 	// peer on any of the channels registered by the reactor
 	Receive(e Envelope)
+
+	// QueueUnprocessedEnvelop is called by the switch when an unprocessed
+	// envelope is received. Unprocessed envelopes are immediately buffered in a
+	// queue to avoid blocking. Incoming messages are then passed to a
+	// processing function. The default processing function unmarshals the
+	// messages in the order the sender sent them and then calls Receive on the
+	// reactor. The queue size and the processing function can be changed by
+	// passing options to the base reactor.
+	QueueUnprocessedEnvelope(e UnprocessedEnvelope)
 }
 
 // --------------------------------------
 
 type BaseReactor struct {
 	service.BaseService // Provides Start, Stop, .Quit
 	Switch              *Switch
+
+	incoming chan UnprocessedEnvelope
+
+	// processor is called with the incoming channel and is responsible for
+	// unmarshalling the messages and calling Receive on the reactor.
+	processor func(incoming <-chan UnprocessedEnvelope) error
 }
 
-func NewBaseReactor(name string, impl Reactor) *BaseReactor {
-	return &BaseReactor{
+type ReactorOptions func(*BaseReactor)
+
+func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReactor {
+	base := &BaseReactor{
 		BaseService: *service.NewBaseService(nil, name, impl),
 		Switch:      nil,
+		incoming:    make(chan UnprocessedEnvelope, 100),
+		processor:   DefaultProcessor(impl),
+	}
+
+	for _, opt := range opts {
+		opt(base)
+	}
+
+	go func() {
+		base.processor(base.incoming)
+	}()
```

Comment:
ahh the reason for not calling `go base.processor(base.incoming)` was that I wasn't sure how to do error handling here. if we want to do the same thing that we are doing with https://github.com/cometbft/cometbft/pull/3209, then we need to stop the peer after catching the panic. 

we still need to fix this before merging. There are few different ways we could handle this, but I haven't found not super hacky way of doing this yet